File: go/grains/grains.go
package grains

import "fmt"

func Square(n int) (uint64, error) {
	var ans uint64 = 1
	if n <= 0 || n > 64 {
		return 0, fmt.Errorf("invalid square: %d", n)
	} else {
		return (ans << uint(n-1)), nil
	}
}

func Total() uint64 {
	return (1 << 64) - 1
}

File: scala/grains/src/main/scala/grains.scala
object Grains extends WithCache[Int, BigInt] {
  private def doSquare(x: Int) = BigInt(1) << (x - 1)

  def square(x: Int) = withCache(x, doSquare)

  def total = (1 to 64).map(square).sum
}

trait WithCache[K,V] {
  private val cache = collection.mutable.Map[K,V]()

  def withCache(key: K, f: K => V) = cache.getOrElseUpdate(key, f(key))
}
